File: LUSIA Studio - Frontend/lib/presentation-generation.ts
/**
 * Presentation Generation — TypeScript types & API client
 *
 * Handles the presentation generation endpoints:
 * - POST /api/presentations/start
 * - GET /api/presentations/{artifactId}/stream (SSE)
 * - GET /api/presentations/{artifactId}
 */

/* ═══════════════════════════════════════════════════════════════
   TYPES
   ═══════════════════════════════════════════════════════════════ */

export interface PresentationStartInput {
    prompt: string;
    size: "short" | "long";
    template: "explicative" | "interactive_explanation" | "step_by_step_exercise";
    subject_id?: string | null;
    year_level?: string | null;
    subject_component?: string | null;
    curriculum_codes: string[];
    upload_artifact_id?: string | null;
}

export interface PresentationStartResult {
    artifact_id: string;
    artifact_name: string;
    artifact_type: string;
    icon: string | null;
    source_type: string;
    subject_id: string | null;
    subject_ids: string[] | null;
    year_level: string | null;
    curriculum_codes: string[] | null;
    is_processed: boolean;
    is_public: boolean;
    created_at: string | null;
}

export interface PresentationPlan {
    title?: string;
    description?: string;
    target_audience?: string;
    total_slides?: number;
    size?: string;
    slides?: Array<{
        id: string;
        phase?: string;
        type?: string;
        subtype?: string | null;
        title?: string;
        intent?: string;
        description?: string;
    }>;
}

export type PresentationStreamEvent =
    | { type: "planning"; message: string }
    | { type: "plan_partial"; plan: PresentationPlan }
    | { type: "plan_complete"; plan: PresentationPlan }
    | { type: "generating_slides"; message: string; total: number }
    | { type: "slide_progress"; current: number; total: number; message: string }
    | { type: "slide_html_snapshot"; slide_id: string; current: number; total: number; html: string }
    | { type: "slide_html_done"; slide_id: string; current: number; total: number; html: string }
    | { type: "done"; artifact_id: string; total_slides?: number }
    | { type: "error"; message: string };


/* ═══════════════════════════════════════════════════════════════
   API FUNCTIONS
   ═══════════════════════════════════════════════════════════════ */

/**
 * Create a presentation artifact and enqueue background generation.
 */
export async function startPresentationGeneration(
    input: PresentationStartInput,
): Promise<PresentationStartResult> {
    const response = await fetch("/api/presentations/start", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(input),
        cache: "no-store",
    });

    if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.detail || err.error || "Failed to start presentation generation");
    }

    const result = await response.json();
    const { cacheInvalidate } = await import("@/lib/cache");
    cacheInvalidate("artifacts:");
    return result;
}

/**
 * Open an SSE stream for presentation generation progress.
 *
 * Uses fetch + ReadableStream (not EventSource) to support auth headers
 * via the Next.js proxy.
 *
 * Returns an AbortController for cleanup on component unmount.
 */
export function streamPresentationGeneration(
    artifactId: string,
    onEvent: (event: PresentationStreamEvent) => void,
    onError: (error: Error) => void,
    onComplete: () => void,
): AbortController {
    const controller = new AbortController();

    (async () => {
        try {
            const response = await fetch(
                `/api/presentations/${artifactId}/stream`,
                {
                    method: "GET",
                    headers: { Accept: "text/event-stream" },
                    signal: controller.signal,
                    cache: "no-store",
                },
            );

            if (!response.ok) {
                throw new Error(`Stream request failed with status ${response.status}`);
            }

            const reader = response.body?.getReader();
            if (!reader) {
                throw new Error("No response body");
            }

            const decoder = new TextDecoder();
            let buffer = "";

            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });

                // Parse SSE events from the buffer
                const lines = buffer.split("\n");
                buffer = lines.pop() || ""; // Keep incomplete line in buffer

                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed || trimmed.startsWith(":")) continue; // Skip empty lines and comments

                    if (trimmed.startsWith("data: ")) {
                        const jsonStr = trimmed.slice(6);
                        try {
                            const event = JSON.parse(jsonStr) as PresentationStreamEvent;
                            onEvent(event);

                            if (event.type === "done" || event.type === "error") {
                                onComplete();
                                return;
                            }
                        } catch {
                            // Skip malformed JSON
                        }
                    }
                }
            }

            onComplete();
        } catch (err) {
            if (controller.signal.aborted) return; // Intentional abort
            onError(err instanceof Error ? err : new Error(String(err)));
        }
    })();

    return controller;
}

/**
 * Fetch a completed presentation artifact.
 */
export async function getPresentation(
    artifactId: string,
): Promise<Record<string, any>> {
    const response = await fetch(`/api/presentations/${artifactId}`, {
        cache: "no-store",
    });

    if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.detail || err.error || "Failed to fetch presentation");
    }

    return response.json();
}

File: LUSIA Studio - Backend/app/api/http/routers/presentation_generation.py
"""
Presentation generation endpoints.

POST /start        — Create artifact + enqueue async generation
GET  /{id}/stream  — SSE endpoint for detailed generation progress (Channel A)
GET  /{id}         — Return completed presentation artifact
"""

import asyncio
import json

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from supabase import Client

from app.api.deps import require_teacher
from app.api.http.schemas.presentation_generation import (
    PresentationStartIn,
    PresentationStartOut,
)
from app.api.http.services.presentation_generation_service import (
    create_presentation_artifact,
    generate_presentation_task,
    get_presentation,
)
from app.core.database import get_b2b_db
from app.pipeline.task_manager import pipeline_manager

router = APIRouter()


def _sse(data: dict) -> str:
    return f"data: {json.dumps(data, ensure_ascii=False)}\n\n"


# ── POST /start ──────────────────────────────────────────────


@router.post("/start", response_model=PresentationStartOut, status_code=201)
async def start_presentation_generation(
    payload: PresentationStartIn,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """Create a presentation artifact and enqueue background generation."""
    org_id = current_user["organization_id"]
    user_id = current_user["id"]

    artifact = create_presentation_artifact(db, org_id, user_id, payload)

    # Enqueue async task — category carries org_id for the custom task_fn
    await pipeline_manager.enqueue(
        artifact_id=artifact["id"],
        job_id=artifact["job_id"],
        user_id=user_id,
        category=org_id,  # Passed as org_id to task_fn
        task_fn=generate_presentation_task,
    )

    return PresentationStartOut(
        artifact_id=artifact["id"],
        artifact_name=artifact["artifact_name"],
        artifact_type="presentation",
        icon=artifact.get("icon"),
        source_type="native",
        subject_id=artifact.get("subject_id"),
        subject_ids=artifact.get("subject_ids"),
        year_level=artifact.get("year_level"),
        curriculum_codes=artifact.get("curriculum_codes"),
        is_processed=False,
        is_public=False,
        created_at=artifact.get("created_at"),
    )


# ── GET /{artifact_id}/stream — Dedicated SSE (Channel A) ───


@router.get("/{artifact_id}/stream")
async def stream_presentation_generation(
    artifact_id: str,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """SSE endpoint for detailed presentation generation progress.

    Subscribes to PipelineTaskManager events for this specific artifact.
    If generation already completed, sends a ``done`` event immediately.
    """
    org_id = current_user["organization_id"]
    user_id = current_user["id"]

    # Check if already completed
    try:
        artifact = get_presentation(db, artifact_id, org_id)
    except Exception:
        raise HTTPException(status_code=404, detail="Presentation not found")

    content = artifact.get("content") or {}
    phase = content.get("phase", "pending")

    async def event_generator():
        # If already done, send immediately
        if phase == "completed":
            plan = content.get("plan", {})
            total_slides = plan.get("total_slides", len(content.get("slides", [])))
            yield _sse({"type": "plan_complete", "plan": plan})
            yield _sse({"type": "done", "artifact_id": artifact_id, "total_slides": total_slides})
            return

        if phase == "failed":
            yield _sse({
                "type": "error",
                "message": artifact.get("processing_error") or "A geração falhou.",
            })
            return

        # Subscribe to real-time updates
        queue = pipeline_manager.subscribe(user_id)
        try:
            # Send current status
            if phase == "planning":
                yield _sse({"type": "planning", "message": "A planear estrutura pedagógica..."})
            elif phase == "generating_slides":
                plan = content.get("plan", {})
                total = plan.get("total_slides", 0)
                yield _sse({"type": "plan_complete", "plan": plan})
                yield _sse({"type": "generating_slides", "message": "A gerar slides...", "total": total})

            # Stream events
            while True:
                try:
                    event = await asyncio.wait_for(queue.get(), timeout=30.0)
                except asyncio.TimeoutError:
                    yield ": keepalive\n\n"
                    continue

                # Only forward events for this artifact
                if event.get("artifact_id") != artifact_id:
                    continue

                event_type = event.get("type")
                step = event.get("step", "")

                if event_type == "status":
                    if step == "planning":
                        yield _sse({"type": "planning", "message": event.get("step_label", "A planear...")})
                    elif step == "generating_slides":
                        # Re-read artifact to get plan
                        try:
                            updated = get_presentation(db, artifact_id, org_id)
                            plan = (updated.get("content") or {}).get("plan", {})
                            total = plan.get("total_slides", 0)
                            yield _sse({"type": "plan_complete", "plan": plan})
                            yield _sse({"type": "generating_slides", "message": "A gerar slides...", "total": total})
                        except Exception:
                            yield _sse({"type": "generating_slides", "message": event.get("step_label", "A gerar slides..."), "total": 0})

                elif event_type == "completed":
                    try:
                        final = get_presentation(db, artifact_id, org_id)
                        final_content = final.get("content") or {}
                        total_slides = len(final_content.get("slides", []))
                        yield _sse({"type": "done", "artifact_id": artifact_id, "total_slides": total_slides})
                    except Exception:
                        yield _sse({"type": "done", "artifact_id": artifact_id})
                    return

                elif event_type == "failed":
                    yield _sse({
                        "type": "error",
                        "message": event.get("error_message", "A geração falhou."),
                    })
                    return

                elif event_type in {"slide_progress", "slide_html_snapshot", "slide_html_done"}:
                    yield _sse(event)

                elif event_type == "plan_partial":
                    yield _sse(event)

        except asyncio.CancelledError:
            pass
        finally:
            pipeline_manager.unsubscribe(user_id, queue)

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


# ── GET /{artifact_id} — Full presentation data ─────────────


@router.get("/{artifact_id}")
async def get_presentation_artifact(
    artifact_id: str,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """Return the full presentation artifact (plan + slides)."""
    org_id = current_user["organization_id"]
    return get_presentation(db, artifact_id, org_id)
